File: dynamic_listings/python/default/collisiondetection/bounding_box.py
from typing import Self


class Point:
    """
    A simple Point
    """

    def __init__(self, x: int, y: int) -> None:
        """
        Creates the point
        :x: The X Coordinate of the point
        :y: The Y Coordinate of the point
        """
        self.x: int = x
        self.y: int = y


class Rectangle:
    """
    A rectangle, made with an upper-left corner, width and height
    """

    def __init__(self, point: Point, width: int, height: int) -> None:
        """Creates the rectange

        :point: The upper left corner point of the rectangle
        :width: The rectangle Width
        :height: The rectangle Height

        """
        self.corner: Point = point
        self.width: int = width
        self.height: int = height

    @staticmethod
    def from_points(topleft: Point, bottomright: Point) -> Self:
        ...


def bounding_box(vertices: list[Point]) -> Rectangle:
    # First we create and bootstrap the variables
    xmin: int = vertices[0].x
    xmax: int = vertices[0].x
    ymin: int = vertices[0].y
    ymax: int = vertices[0].y
    # Now we iterate through all the other vertices
    for vertex in vertices:
        xmin = min(xmin, vertex.x)
        xmax = max(xmax, vertex.x)
        ymin = min(ymin, vertex.y)
        ymax = max(ymax, vertex.y)
    # Now we can build the needed points for the bounding box
    A: Point = Point(xmin, ymin)
    C: Point = Point(xmax, ymax)
    # We build our bounding box
    boundingBox: Rectangle = Rectangle.from_points(A, C)
    # and return it
    return boundingBox

File: dynamic_listings/rust/default/collisiondetection/bounding_box.rs
struct Point {
    // Rewritten as a memo
    x: u32,
    y: u32,
}

struct Rectangle {
    corner: Point,
    width: u32,
    height: u32,
    // ...
}

impl Rectangle {
    // ...
    fn from_points(topleft: &Point, bottomright: &Point) -> Self {
        // ...
    }

    fn bounding_box(vertices: &[Point]) -> Self {
        // First we create and bootstrap the variables
        let mut xmin = vertices[0].x;
        let mut xmax = vertices[0].x;
        let mut ymin = vertices[0].y;
        let mut ymax = vertices[0].y;
        // Now we iterate through all the other vertices
        for vertex in vertices {
            if vertex.x < xmin {
                xmin = vertex.x;
            }
            if vertex.x > xmax {
                xmax = vertex.x;
            }
            if vertex.y < ymin {
                ymin = vertex.y;
            }
            if vertex.y > ymax {
                ymax = vertex.y;
            }
        }
        // Now we can build the needed points for the bounding box
        let a = Point { x: xmin, y: ymin };
        let c = Point { x: xmax, y: ymax };
        // We build our bounding box
        Self::from_points(&a, &c)
    }
}
